teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,354 @@
+an authenticated `git` clone of a private repo may require that a `.ssh` directory
+Workspace be provided by a TaskRun but cloning public repos could occur without it.
+
+## Motivation
+
+Several broad areas motivate this proposal:
+
+1. Credentials
+2. Supplying optional data to inform a Task's behaviour
+3. Receiving optional data generated by a Task
+
+### Credentials
+
+Allow Tasks to declare dependence / support for specific optional credentials
+mounted as files. Example: a .ssh directory with private key and known_hosts.
+
+Optionality is desirable because:
+1. Multiple credential formats are often supported. E.g. git supporting both
+.ssh directory or a .gitconfig file. These could be modeled as two separate
+optional Workspaces.
+2. Credentials can sometimes be safely omitted. E.g. git does not need
+credentials to clone a public repo like Tekton Pipelines.
+3. Tekton currently has multiple competing mechanisms for mounting files,
+specifically credential files, into a Task: creds-init, Workspaces,
+volumes/volumeMounts. Writing a Catalog Task to support multiple mechanisms
+implies each can be optionally omitted.
+
+### Supplying optional data to inform a Task's behaviour
+
+Allow Tasks to accept an optional set of data that will change the operation
+of that Task in some way.
+
+Examples of this include:
+- accepting an optional linter configuration to change the warnings/errors that are
+raised from the defaults.
+- accepting an optional kubeconfig file that allows a Task to target an external
+cluster instead of the same cluster it's running in.
+- accepting an optional manifest naming specific files to be uploaded instead of
+simply uploading an entire directory of content.
+
+### Receiving optional data generated by a Task
+
+Allow Tasks to _optionally_ generate some data and supply that to the user
+or other Tasks in a Pipeline.
+
+Examples of usage include:
+- Receiving test results as a file as well as printed to stdout.
+- Receiving a manifest of file paths that were copied to external storage.
+- Receiving a diff of files that were modified during Task execution.
+
+All of the above are examples of behaviours that could be opt-in: a user/pipeline
+may not need the results of a test run beyond a pass/fail exit code. A team may
+not need to care which specific files were copied out to external storage; only
+that the transfer succeeded. The diff of modifications made during execution may
+have no bearing on subsequent Tasks in a Pipeline.
+
+A Task can also conceivably skip some portion of work if the user has elected
+not to receive the generated data. For example, a Task may not need to compute
+the diff of its modifications or generate a manifest of changed files if the
+user has not provided a location for that data to be received.
+
+### Goals
+
+1. Allow Tasks to declare optional support for file-based credentials.
+2. Enable Tasks to offer support of Tekton's multiple competing credential mechanisms.
```

Comment:
Isn't this *just* a benefit we get from having optional support for workspace ? (instead of a goal)

Reply:
From the perspective of a catalog task author I actually think this is a desirable goal which should be used to decide if the feature is successful or not.  If we got to the end of this TEP and catalog authors were stuck writing separate tasks to support workspaces and creds-init I'd consider that a failure of the design / tep.